external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -165,16 +183,112 @@ class AvroSerializer(rootCatalystType: DataType, rootAvroType: Schema, nullable:
+  // type Schema.Type.LONG
+  private def resolveNullableType(avroType: Schema, catalystType: DataType,
+                                  nullable: Boolean): Schema = {
+    (nullable, avroType.getType) match {
+      case (false, Type.UNION) | (true, Type.UNION) =>
+        // avro uses union to represent nullable type.
+        val fieldTypes = avroType.getTypes.asScala
+
+        // If we're nullable, we need to have at least two types.  Cases with more than two types
+        // are captured in test("read read-write, read-write w/ schema, read") w/ test.avro input
+        assert(fieldTypes.length >= 2)
+
+        val actualType = catalystType match {
+          case NullType => fieldTypes.filter(_.getType == Type.NULL)
+          case BooleanType => fieldTypes.filter(_.getType == Type.BOOLEAN)
+          case ByteType => fieldTypes.filter(_.getType == Type.INT)
+          case BinaryType =>
+            val at = fieldTypes.filter(x => x.getType == Type.BYTES || x.getType == Type.FIXED)
+            if (at.length > 1) {
+              throw new IncompatibleSchemaException(
+                s"Cannot resolve schema of ${catalystType} against union ${avroType.toString}")
+            } else {
+              at
+            }
+          case ShortType | IntegerType => fieldTypes.filter(_.getType == Type.INT)
+          case LongType => fieldTypes.filter(_.getType == Type.LONG)
+          case FloatType => fieldTypes.filter(_.getType == Type.FLOAT)
+          case DoubleType => fieldTypes.filter(_.getType == Type.DOUBLE)
+          case d: DecimalType => fieldTypes.filter(_.getType == Type.STRING)
+          case StringType => fieldTypes
+            .filter(x => x.getType == Type.STRING || x.getType == Type.ENUM)
+          case DateType => fieldTypes.filter(x => x.getType == Type.INT || x.getType == Type.LONG)
+          case TimestampType => fieldTypes.filter(_.getType == Type.LONG)
+          case ArrayType(et, containsNull) =>
+            // Find array that matches the element type specified
+            fieldTypes.filter(x => x.getType == Type.ARRAY
+              && typeMatchesSchema(et, x.getElementType))
+          case st: StructType => // Find the matching record!
+            val recordTypes = fieldTypes.filter(x => x.getType == Type.RECORD)
+            if (recordTypes.length > 1) {
+              throw new IncompatibleSchemaException(
+                "Unions of multiple record types are NOT supported with user-specified schema")
+            }
+            recordTypes
+          case MapType(kt, vt, valueContainsNull) =>
+            // Find the map that matches the value type.  Maps in Avro are always key type string
+            fieldTypes.filter(x => x.getType == Type.MAP && typeMatchesSchema(vt, x.getValueType))
+          case other =>
+            throw new IncompatibleSchemaException(s"Unexpected type: $other")
+        }
+
+        assert(actualType.length == 1)
```

Comment:
We need to show error message if the length is not 1.

Reply:
Can you elaborate when  `actualType.length == 0` or `actualType.length > 1`?

Is it possible that `catalystType` is `Int`, and `fieldTypes` only contains `Long`? Do we want to do the promotion?

Reply:
I've removed ALL asserts from the code and replaced them with clearer exceptions.